### PR TITLE
Removed Site Template Contract Service Graph logic from mso_schema_template_contract_service_graph resource

### PR DESCRIPTION
**Pull request by sajagana**
Note:

Removed Site Template Contract Service Graph logic from mso_schema_template_contract_service_graph resource. PR belongs to: https://github.com/CiscoDevNet/terraform-provider-mso/issues/190

### CODE REVIEW

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 1]
```diff
@@ -90,608 +72,176 @@ func resourceTemplateContractServiceGraph() *schema.Resource {
...
-		return nil, fmt.Errorf("No service graph found for given name")
+func resourceMSOSchemaTemplateContractServiceGraphRead(d *schema.ResourceData, m interface{}) error {
+	log.Printf("[DEBUG] Begining Read Template Contract Service Graph")
+	serviceGraphTokens := strings.Split(d.Id(), "/")
+	d.Set("schema_id", serviceGraphTokens[1])
+	d.Set("template_name", serviceGraphTokens[3])
+	d.Set("contract_name", serviceGraphTokens[5])
+	err := setSchemaTemplateContractServiceGraphAttrs(d, m, false)
+	if err != nil {
+		return err
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:159]: do we need the errorForObjectNotFound function here?

**Review by akinross — CHANGES_REQUESTED** [review]

File: mso/resource_mso_schema_template_contract_service_graph.go (hunk 1 above)
Comment (sajagana) [reply mso/resource_mso_schema_template_contract_service_graph.go:159]: errorForObjectNotFound used in the setSchemaTemplateContractServiceGraphAttrs function.

Comment (akinross) [reply mso/resource_mso_schema_template_contract_service_graph.go:159]: I did a search and in all other resources and data sources this is only called from the read function. Should we keep this consistent to ensure we are not introducing any undesired behaviour?

Comment (sajagana) [reply mso/resource_mso_schema_template_contract_service_graph.go:159]: Sure I will move back to the read function.

File: examples/schema_template_contract_service_graph/main.tf [hunk 2]
```diff
@@ -13,39 +13,66 @@ provider "mso" {
...
-    consumer_connector_bd_template_name = "Template1"
-    provider_connector_cluster_interface = "test"
-    consumer_connector_cluster_interface = "test"
-    provider_connector_redirect_policy_tenant = "NkAutomation"
-    provider_connector_redirect_policy = "test2"
-    consumer_connector_redirect_policy_tenant = "NkAutomation"
-    consumer_connector_redirect_policy = "test2"
-    provider_subnet_ips = ["1.2.3.4/20"]
-    consumer_subnet_ips = ["1.2.3.4/20"]
+    consumer_connector_bd_name = data.mso_schema_template_bd.bd1.name
```
Comment (shrsr) [inline examples/schema_template_contract_service_graph/main.tf:71]: Can both  consumer_connector_bd_name and provider_connector_bd_name have the same value?

File: examples/schema_template_contract_service_graph/main.tf [hunk 3]
```diff
@@ -13,39 +13,66 @@ provider "mso" {
...
+    provider_connector_bd_name = data.mso_schema_template_bd.bd1.name
   }
-}
\ No newline at end of file
+  node_relationship {
+    consumer_connector_bd_name = data.mso_schema_template_bd.bd1.name
```
Comment (shrsr) [inline examples/schema_template_contract_service_graph/main.tf:75]: Can both  consumer_connector_bd_name and provider_connector_bd_name have the same value?

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 4]
```diff
@@ -30,56 +30,38 @@ func resourceTemplateContractServiceGraph() *schema.Resource {
...
 			"service_graph_template_name": &schema.Schema{
 				Type:         schema.TypeString,
-				Optional:     true,
 				ForceNew:     true,
+				Optional:     true,
+				Computed:     true,
 				ValidateFunc: validation.StringLenBetween(1, 1000),
 			},
-
 			"node_relationship": &schema.Schema{
```
Comment (shrsr) [inline mso/resource_mso_schema_template_contract_service_graph.go:65]: Does the order of the node_relationship blocks matter?

File: website/docs/r/schema_template_contract_service_graph.html.markdown [hunk 5]
```diff
@@ -3,98 +3,63 @@ layout: "mso"
...
+  * `consumer_connector_bd_template_name` - (Optional) The template name of the BD that has to be connected to a Consumer Connector. The `template_name` will be used if not provided.
 
 ## Attribute Reference ##
 
 No attributes are exported.
 
 ## Importing ##
 
-An existing MSO Schema Template Contract service graph can be [imported][docs-import] into this resource via its Id/path, via the following command: [docs-import]: <https://www.terraform.io/docs/import/index.html>
+An existing MSO Schema Template Contract Service Graph can be [imported][docs-import] into this resource via its Id/path, via the following command: [docs-import]: <https://www.terraform.io/docs/import/index.html>
```
Comment (shrsr) [inline website/docs/r/schema_template_contract_service_graph.html.markdown:61]: An existing MSO Schema Template Contract Service Graph can be [imported][docs-import] into this resource using its Id/path, via the following command: [docs-import]: <https://www.terraform.io/docs/import/index.html>

File: mso/resource_mso_schema_template_contract_service_graph.go (hunk 4 above)
Comment (shrsr) [reply mso/resource_mso_schema_template_contract_service_graph.go:65]: If yes, lets document it

**Review by shrsr — CHANGES_REQUESTED** [review]

File: examples/schema_template_contract_service_graph/main.tf (hunk 3 above)
Comment (sajagana) [reply examples/schema_template_contract_service_graph/main.tf:75]: Yes, it accepts.

**Review by shrsr — DISMISSED** [review]
LGTM

**Review by akinross — DISMISSED** [review]
LGTM

**Review by samiib — DISMISSED** [review]
LGTM 👍

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 6]
```diff
@@ -1085,229 +392,83 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
+// PostSchemaTemplateContractServiceGraphConfig creates a new service graph configuration for a template contract.
+//
+// Parameters:
+//   - action: The action to perform.
+//   - d: The schema resource data.
+//   - m: The client interface.
+//
+// Returns:
+//   - An error if there was a problem creating the service graph configuration.
+func PostSchemaTemplateContractServiceGraphConfig(action string, d *schema.ResourceData, m interface{}) error {
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:415]: why named action? perhaps it is more clear to keep consistent with go-client naming? Also if you write the function docs, you might want to specify the possible actions?

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 7]
```diff
@@ -1085,229 +392,83 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
 	msoClient := m.(*client.Client)
-
-	schemaID := d.Get("schema_id").(string)
-	siteID := d.Get("site_id").(string)
+	schemaId := d.Get("schema_id").(string)
 	TemplateName := d.Get("template_name").(string)
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:418]:
```suggestion
	templateName := d.Get("template_name").(string)
```

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 8]
```diff
@@ -1085,229 +392,83 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
 	contractName := d.Get("contract_name").(string)
+	serviceGraph := d.Get("service_graph_name").(string)
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:420]: should be keep the naming aligned with the schema definition?
```suggestion
	serviceGraphName := d.Get("service_graph_name").(string)
```

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 9]
```diff
@@ -1085,229 +392,83 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
-	return nil
-}
+	serviceGraphRef := make(map[string]interface{})
+	serviceGraphRef["serviceGraphName"] = serviceGraph
 
-func getTemplateServiceGraph(cont *container.Container, templateName, graphName string) (*container.Container, int, error) {
-	templateCount, err := cont.ArrayCount("templates")
-	if err != nil {
-		return nil, -1, fmt.Errorf("No Template found")
+	if graphSchema, ok := d.GetOk("service_graph_schema_id"); ok {
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:425]: should be keep the naming aligned with the schema definition?
```suggestion
	if serviceGraphSchemaId, ok := d.GetOk("service_graph_schema_id"); ok {
```

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 10]
```diff
@@ -1085,229 +392,83 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
-
-				apiSgName := models.StripQuotes(sgCont.S("name").String())
-
-				if apiSgName == graphName {
-					return sgCont, j, nil
-				}
-			}
-
-		}
+	if graphTemp, ok := d.GetOk("service_graph_template_name"); ok {
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:431]: should be keep the naming aligned with the schema definition?
```suggestion
	if serviceGraphTemplateName, ok := d.GetOk("service_graph_template_name"); ok {
```

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 11]
```diff
@@ -1085,229 +392,83 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
-			return nil, -1, fmt.Errorf("Unable to site service node element")
-		}
 
-		nodeRef := models.StripQuotes(nodeCont.S("serviceNodeRef").String())
-
-		nodeSplit := strings.Split(nodeRef, "/")
-		if len(nodeSplit) == 9 {
-			if nodeSplit[2] == schemaId && nodeSplit[4] == templateName && nodeSplit[6] == graphName && nodeSplit[8] == nodeName {
-				return nodeCont, i, nil
+	tempNodeList := d.Get("node_relationship").([]interface{})
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:452]: should we rename "tempNodeList" and "nodeList" to better distinguish difference between from schema input and from ndo node configuration?

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 12]
```diff
@@ -892,49 +291,54 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
-							nodeInterface := d.Get("node_relationship")
 
 							probdRef := models.StripQuotes(node.S("providerConnector", "bdRef").String())
 							probdRefTokens := strings.Split(probdRef, "/")
 							relationMap["provider_connector_bd_name"] = probdRefTokens[len(probdRefTokens)-1]
 
-							if checkNodeAttr(nodeInterface, "provider_connector_bd_schema_id", k) {
+							if importFlag {
+								relationMap["provider_connector_bd_schema_id"] = probdRefTokens[len(probdRefTokens)-5]
+							} else if checkNodeAttr(nodeInterface, "provider_connector_bd_schema_id", k) {
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:333]: shoudl we combine the conditional?
```suggestion
							if importFlag || checkNodeAttr(nodeInterface, "provider_connector_bd_schema_id", k) {
```

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 13]
```diff
@@ -733,137 +244,25 @@ func resourceTemplateContractServiceGraphUpdate(d *schema.ResourceData, m interf
...
+// for a template contract in the schema.
+//
+// Parameters:
+// - d: *schema.ResourceData: The resource data containing the schema attributes.
+// - m: interface{}: The client interface.
+// - importFlag: bool: The import flag to determine if the attributes should be imported.
+//
+// Returns:
+// - error: The error encountered, if any.
+func setSchemaTemplateContractServiceGraphAttrs(cont *container.Container, d *schema.ResourceData, importFlag bool) error {
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:261]: What is the value of this import_flag? I understand that it is to differentiate between import and read, but should those not do the same thing?

File: mso/resource_mso_schema_template_contract_service_graph.go [hunk 14]
```diff
@@ -969,113 +378,11 @@ func resourceTemplateContractServiceGraphRead(d *schema.ResourceData, m interfac
...
-				}
-			}
-		}
-		if foundSite {
-			break
-		}
-	}
-
-	if foundSite && foundTemp {
+	if foundTemp {
```
Comment (akinross) [inline mso/resource_mso_schema_template_contract_service_graph.go:381]: why this flag and not execute the code in the location of foundTemp (line 370)?

**Review by akinross — CHANGES_REQUESTED** [review]

File: mso/resource_mso_schema_template_contract_service_graph.go (hunk 13 above)
Comment (sajagana) [reply mso/resource_mso_schema_template_contract_service_graph.go:259]: The import_flag is used to ignore the resource attributes during the terraform plan.

**Review by samiib — DISMISSED** [review]
LGTM

**Review by akinross — APPROVED** [review]
LGTM

File: website/docs/r/schema_template_contract_service_graph.html.markdown [hunk 15]
```diff
@@ -3,98 +3,66 @@ layout: "mso"
...
+  Manages MSO Schema Template Contract Service Graph.
 ---
 
 # mso_schema_template_contract_service_graph #
 
-Manages MSO Schema Template Contract service graph.
+Manages MSO Schema Template Contract Service Graph.
+
+# Note: #
+Only MSO versions 3.7 and 4.2+ are compatible with this resource. For the MSO version 4.1 is not supported.
```
Comment (lhercot) [inline website/docs/r/schema_template_contract_service_graph.html.markdown:14]:
```suggestion
This resource is only compatible with NDO versions 3.7 and 4.2+. NDO versions 4.0 and 4.1 are not supported.
```

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**